### ISSUE TITLE
shmget limited to 32 bit segment size on 64 bit OS

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9685]
Migrated from [rt.perl.org#63924](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=63924) (status was 'resolved')

Searchable as RT63924$

**Comment by p5pRT** [comment]
### From @mhx

#### Created by mhx@cpan.org



This was originally reported as cpan \#44182 for IPC&#8203;::SysV\. But the
underlying problem is that the perl core doesn't support more than
31 bits in the size argument to shmget\(\)\, as n in Perl\_do\_ipcget\(\)
is declared as I32\. However\, the size parameter of shmget\(\) is of
type size\_t\, which can be larger \(and is unsigned\)\.

Quoting the original bug report&#8203;:

> I installed IPC\-SysV 2\.0\.0 on CentOs 5\.1 x86\_x64 bit version\.
> Perl version is v5\.8\.8
> 
> I tested with "shmsizetest\.pl" script\, varing $shmsize from 1G~4G
> 
> It was successful to create 1 G byte shared memory\,
> but 2G\, 3G\, 4G were failed\. \(1G=1x1024x1024x1024\)
> And Errno\(=$\!\) was "Invalide Argument"
> Parameters related to SysV shared memory are
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> max number of segments = 4096
> max seg size \(kbytes\) = 67108864
> max total shared memory \(kbytes\) = 17179869184
> min seg size \(bytes\) = 1
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> So I think no problem to get 4G byte at SysV shared memory
> \(C was successful to create 4G shared memory\, using shmget function\)
> 
> I wander IPC&#8203;::SharedMem is not supporting 64 bit os?
> 
> Actually is the argument $SIZE of IPC&#8203;::SharedMem limited as 4byte 
> signed integer?

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.11.0:

Configured by mhx at Sun Mar 15 04:46:39 MET 2009.

Summary of my perl5 (revision 5 version 11 subversion 0 patch GitLive-blead-774-g78b3516) configuration:
  Snapshot of: 78b3516b6c02d638c0127c4ce6ebe476158c5545
  Platform:
    osname=linux, osvers=2.6.26.7mhx, archname=i686-linux-thread-multi-64int-ld
    uname='linux r2d2 2.6.26.7mhx #1 smp thu nov 27 21:45:03 met 2008 i686 intel(r) core(tm)2 cpu t7400 @ 2.16ghz genuineintel gnulinux '
    config_args='-des -Uinstallusrbinperl -Dusethreads -Duse64bitint -Duselongdouble -Dcf_email=mhx@cpan.org -Dcc=gcc -DDEBUGGING=both -Doptimize=-ggdb3 -Dprefix=/home/mhx/perl/blead-debug -Dusedevel -Uversiononly -Aappend:ccflags= -DPERL_GLOBAL_STRUCT_PRIVATE -DPL_OP_SLAB_ALLOC -W -Wall '
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=undef, uselongdouble=define
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DPERL_GLOBAL_STRUCT_PRIVATE -DPL_OP_SLAB_ALLOC -W -Wall -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-ggdb3',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DPERL_GLOBAL_STRUCT_PRIVATE -DPL_OP_SLAB_ALLOC -W -Wall -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector'
    ccversion='', gccversion='4.3.2', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long long', ivsize=8, nvtype='long double', nvsize=12, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldb -ldl -lm -lcrypt -lutil -lpthread -lc -lgdbm_compat
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=/lib/libc-2.6.1.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.6.1'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -ggdb3 -L/usr/local/lib -fstack-protector'

Locally applied patches:
    PERL_GIT_UNPUSHED_COMMITS    	/* do not remove this line */
    PERL_GIT_UNCOMMITTED_CHANGES	/* do not remove this line */


@INC for perl 5.11.0:
    /home/m5/sw/tool/src_general/tparser/lib
    /home/mhx/perl/blead-debug/lib/site_perl/5.11.0/i686-linux-thread-multi-64int-ld
    /home/mhx/perl/blead-debug/lib/site_perl/5.11.0
    /home/mhx/perl/blead-debug/lib/5.11.0/i686-linux-thread-multi-64int-ld
    /home/mhx/perl/blead-debug/lib/5.11.0
    /home/mhx/perl/blead-debug/lib/site_perl
    .


Environment for perl 5.11.0:
    HOME=/home/mhx
    LANG=de_DE.utf8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/trolltech/qt/bin:/home/mhx/apps/bin/ccache:/usr/bin/ccache:/home/mhx/apps/bin:/usr/local/bin:/usr/bin:/bin:/opt/bin:/usr/i686-pc-linux-gnu/avr/gcc-bin/4.3.2:/usr/i686-pc-linux-gnu/gcc-bin/4.3.2:/opt/intel/cc/10.1.018/bin:/opt/blackdown-jdk-1.4.2.03/bin:/opt/blackdown-jdk-1.4.2.03/jre/bin:/usr/kde/3.5/bin:/usr/qt/3/bin:/usr/games/bin:/opt/vmware/workstation/bin:/opt/SnapshotCM/bin:/home/m5/sw/tool/bin:/users/patient_data/tool/bin/scripts
    PERL5LIB=/home/m5/sw/tool/src_general/tparser/lib
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
@mhx - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@mhx - Status changed from 'open' to 'new'

**Comment by p5pRT** [comment]
@mhx - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@mhx - Status changed from 'open' to 'new'

**Comment by p5pRT** [comment]
### From @mhx

Attaching test script from original report\.

**Comment by p5pRT** [comment]
### From @mhx

[shmsizetest.pl](https://rt-archive.perl.org/perl5/Ticket/Attachment/541766/257168/shmsizetest.pl)

**Comment by p5pRT** [comment]
### From @mhx

Could you please try to build the development version of perl
with the attached patch applied on top of it and see if that
fixes your problem? I don't have a system where I could test
this right now\. You can get the development version either
by git or by rsync\. For details\, see&#8203;:

&nbsp;  http&#8203;://perl5\.git\.perl\.org/perl\.git/blob/HEAD&#8203;:/pod/perlhack\.pod

Thanks\,
Marcus

**Comment by p5pRT** [comment]
### From @mhx

<details><summary>doio.c.diff</summary>

```diff
--- doio.c.orig	2009-03-16 10:25:55.000000000 +0100
+++ doio.c	2009-03-16 10:28:46.000000000 +0100
@@ -1981,7 +1981,7 @@
 {
     dVAR;
     const key_t key = (key_t)SvNVx(*++mark);
-    const I32 n = (optype == OP_MSGGET) ? 0 : SvIVx(*++mark);
+    SV *nsv = optype == OP_MSGGET ? NULL : *++mark;
     const I32 flags = SvIVx(*++mark);
 
     PERL_ARGS_ASSERT_DO_IPCGET;
@@ -1996,11 +1996,11 @@
 #endif
 #ifdef HAS_SEM
     case OP_SEMGET:
-	return semget(key, n, flags);
+	return semget(key, (int) SvIV(nsv), flags);
 #endif
 #ifdef HAS_SHM
     case OP_SHMGET:
-	return shmget(key, n, flags);
+	return shmget(key, (size_t) SvUV(nsv), flags);
 #endif
 #if !defined(HAS_MSG) || !defined(HAS_SEM) || !defined(HAS_SHM)
     default:

```
</details>

**Comment by p5pRT** [comment]
@mhx - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From ssgo@strastar.com


Dear Marcus
 
I had compiled perl development version 5\.11 with patch file which you gave
me\,

and test again with "shmsizetest\.pl" script which make more than 2G byte
shared memory on 64bit OS\.

It was successfull to create 20G byte with default configuration of SysV on
CentOS x86\_64 \(RedHat version 5\.1 x64\)

Thanks your patch

Best regards
Sangsun Go

\-\-\-\-\-Original Message\-\-\-\-\-
From&#8203;: Marcus Holland\-Moritz via RT \[mailto&#8203;:perlbug\-followup@&#8203;perl\.org\] 
Sent&#8203;: Monday\, March 16\, 2009 7&#8203;:01 PM
Cc&#8203;: ssgo@&#8203;strastar\.com
Subject&#8203;: \[perl \#63924\] shmget limited to 32 bit segment size on 64 bit OS 

Could you please try to build the development version of perl with the
attached patch applied on top of it and see if that fixes your problem? I
don't have a system where I could test this right now\. You can get the
development version either by git or by rsync\. For details\, see&#8203;:

&nbsp;  http&#8203;://perl5\.git\.perl\.org/perl\.git/blob/HEAD&#8203;:/pod/perlhack\.pod

Thanks\,
Marcus

**Comment by p5pRT** [comment]
### From ssgo@strastar.com

[shmsizetest.pl](https://rt-archive.perl.org/perl5/Ticket/Attachment/544846/258790/shmsizetest.pl)

**Comment by p5pRT** [comment]
### From @mhx

Fixed by the following change&#8203;:

http&#8203;://perl5\.git\.perl\.org/perl\.git/commitdiff/c331296618a8003690577e0901fd07183a76094e

**Comment by p5pRT** [comment]
@mhx - Status changed from 'open' to 'resolved'
